# Assignment 1 - AI 2

For this exercise we are going to use the **Bank Marketing dataset** from Moro et al. , which tries to 1
predict if a marketing initiative of a banking institution will results in the clients subscribing into a
service.

The dataset can be downloaded from the UCI repository (https://archive.ics.uci.edu/ml/datasets/
Bank+Marketing), with the bank.csv file being the one that we are going to use.

This dataset consists on 15 features (all of them explained in detail in the dataset), and one binary
output variable ('y') which indicates if the consumer has subscribed or not to the service. Beware
that not all the features are encoded as numeric values, and the target feature is also encoded as
a categorical variable.

In [95]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Exercise 1

**Parse the dataset and obtain the target column as a separate variable.**

**(Hint) Check the delimiter of the dataset, and use var.drop() and var['lab'] to assign the values**

In [96]:
data = pd.read_csv('C:/Users/Asus/Desktop/Python Projects/AI 2 Course/bank.csv', delimiter=';')
data.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

I used the read_csv method with delimiter parameter set to ; to be in accordance with the csv file.

In [97]:
target = data['y']
target.head()

0    no
1    no
2    no
3    no
4    no
Name: y, dtype: object

In [98]:
data.drop('y',axis=1, inplace=True)
data.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome  
0  cellular   19   oct        79         1     -1         0  unknown  
1  cellular   11   may       220         1    339         4  failure  
2  cellular   16   apr       185         1    330         1  failure  
3   unknown    3   jun       199         4     -1         0  unknown  
4   unknown    5   may       226         1     -1         0  unknown

### Exercise 2

**Obtain a report of the different features inside the dataset, check if all the features are useful for our prediction, if not drop them.**

In question 2, the only thing that it is asked is for you to visualize the amount of data for each one of the categories that you have (categorical variables only and using native pandas functions). Once you have the amount of data in each category, and using the description of each feature provided in the dataset (UCI webpage) try to establish if some of the features my introduce target leakage and can be dropped from the dataset (always with a justified reasoning) or if due to certain features only containing a very small number data from a certain category can introduce an imbalance to the dataset (which can have a negative influence in the accuracy numbers prior to validation).


In [99]:
data

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome  
0     cellular   19   oct        79         1     -1         0  unknown  
1     cellular   11   may       220         1    339         4  failure  
2     cellular   16   apr       185         1    330         1  failure  
3      unknown    3   jun       199         4     -1         0  unknown  
4      unknown    5   may       226         1     -1         0  unknown  
...        ...  ...   ...       ...       ...    ...       ...      ...  
4516  cellular   30   jul       329         5     -1         0  unknown  
4517   unknown    9   may       153         1     -1         0  unknown  
4518  cellular   19   aug       151        11     -1         0  unknown  
4519  cellular    6   feb       129         4    211         3    other  
4520  cellular    3   apr       345         2    249         7    other  

[4521 rows x 16 columns]

**Job Variable**

In [100]:
data['job'].value_counts()

management       969
blue-collar      946
technician       768
admin.           478
services         417
retired          230
self-employed    183
entrepreneur     168
unemployed       128
housemaid        112
student           84
unknown           38
Name: job, dtype: int64

**Marital Variable**

In [101]:
data['marital'].value_counts()

married     2797
single      1196
divorced     528
Name: marital, dtype: int64

**Education Variable**

In [102]:
data['education'].value_counts()

secondary    2306
tertiary     1350
primary       678
unknown       187
Name: education, dtype: int64

**Default Variable**

In [103]:
data['default'].value_counts()

no     4445
yes      76
Name: default, dtype: int64

**Housing Variable**

In [104]:
data['housing'].value_counts()

yes    2559
no     1962
Name: housing, dtype: int64

**Loan Variable**

In [105]:
data['loan'].value_counts()

no     3830
yes     691
Name: loan, dtype: int64

**Contact Variable**

In [106]:
data['contact'].value_counts()

cellular     2896
unknown      1324
telephone     301
Name: contact, dtype: int64

We can see that there are a lot of missing values in the contact variable and since it doesn't seem to be a good indicator of the campaign's outcome it is better to drop it:

In [107]:
data.drop('contact', axis=1, inplace=True)

**Month Variable**

In [108]:
data['month'].value_counts()

may    1398
jul     706
aug     633
jun     531
nov     389
apr     293
feb     222
jan     148
oct      80
sep      52
mar      49
dec      20
Name: month, dtype: int64

**Past outcome Variable**

In [109]:
data['poutcome'].value_counts()

unknown    3705
failure     490
other       197
success     129
Name: poutcome, dtype: int64

Here, we have a lot of missing values but the variable, when known, seems to be a very good indicator of our target.

##### Numerical Variables

In [110]:
data.describe()

age       balance          day     duration     campaign  \
count  4521.000000   4521.000000  4521.000000  4521.000000  4521.000000   
mean     41.170095   1422.657819    15.915284   263.961292     2.793630   
std      10.576211   3009.638142     8.247667   259.856633     3.109807   
min      19.000000  -3313.000000     1.000000     4.000000     1.000000   
25%      33.000000     69.000000     9.000000   104.000000     1.000000   
50%      39.000000    444.000000    16.000000   185.000000     2.000000   
75%      49.000000   1480.000000    21.000000   329.000000     3.000000   
max      87.000000  71188.000000    31.000000  3025.000000    50.000000   

             pdays     previous  
count  4521.000000  4521.000000  
mean     39.766645     0.542579  
std     100.121124     1.693562  
min      -1.000000     0.000000  
25%      -1.000000     0.000000  
50%      -1.000000     0.000000  
75%      -1.000000     0.000000  
max     871.000000    25.000000

###### Keep

__*age*__ should be useful because behaviour patterns tend to change with it. __*balance*__ also seems interesting because it is a proxy for loyalty to the bank and financial situation.

__*day*__ might be important because it is a proxy of how much money people have to spend for the rest of the month and might influence their committment to the campaign.

__*pdays*__ seems not to significant for our prediction but we can leave it for now and reassess after the first simple model.

__*campaign*__ might be interesting for our analysis because it is tied to customer loyalty and could help identify a the outcome.


###### Drop

__*previous*__ shouldn't be too important, especially knowing the the last outcome so we can drop it. (poutcome)

__*duration*__ has to be dropped because it introduces target leakage.

In [111]:
data.drop(['duration', 'previous'], axis=1, inplace=True)

In [112]:
data.head()

age          job  marital  education default  balance housing loan  day  \
0   30   unemployed  married    primary      no     1787      no   no   19   
1   33     services  married  secondary      no     4789     yes  yes   11   
2   35   management   single   tertiary      no     1350     yes   no   16   
3   30   management  married   tertiary      no     1476     yes  yes    3   
4   59  blue-collar  married  secondary      no        0     yes   no    5   

  month  campaign  pdays poutcome  
0   oct         1     -1  unknown  
1   may         1    339  failure  
2   apr         1    330  failure  
3   jun         4     -1  unknown  
4   may         1     -1  unknown

We now have 13 variables that we're going to use for the classification problem. 5 of them are numerical and the other 8 are categorical.

### Exercise 3

**As most of the features are of a categorical nature, apply the LabelEncoder to each one of the features.**

As some of you have reached out regarding question 3 of the assignment, and in order to prevent further confusion with the topic, I have decided to change question 3 so only the data of the target column 'y' is encoded with the LabelEncoder, and the rest of the categorical data is encoded with the OrdinalEncoder. Regarding this change, you will find an updated pdf on the assignment and attached to this mail/post. Also, as probably some of you have already performed the assignment, I am giving all of you an extra day so you can accomplish all this modifications.

In [113]:
from sklearn.preprocessing import OrdinalEncoder, LabelEncoder

Apply Ordinal Encoder to categorical variables __*job, marital, education, default, housing, loan, month and poutcome*__

In [114]:
enc = OrdinalEncoder().fit(data[['job','marital','education','default','housing','loan','month','poutcome']])

In [115]:
data[['job','marital','education','default','housing','loan','month','poutcome']] = enc.transform(data[['job','marital','education','default','housing','loan','month','poutcome']])

Also, Label Encode the target variable: __*y*__

In [116]:
enc_l = LabelEncoder()
enc_l.fit(target)
target = enc_l.transform(target)

In [117]:
target.dtype

dtype('int32')

**Updated Dataset**

In [118]:
data.head()

age   job  marital  education  default  balance  housing  loan  day  month  \
0   30  10.0      1.0        0.0      0.0     1787      0.0   0.0   19   10.0   
1   33   7.0      1.0        1.0      0.0     4789      1.0   1.0   11    8.0   
2   35   4.0      2.0        2.0      0.0     1350      1.0   0.0   16    0.0   
3   30   4.0      1.0        2.0      0.0     1476      1.0   1.0    3    6.0   
4   59   1.0      1.0        1.0      0.0        0      1.0   0.0    5    8.0   

   campaign  pdays  poutcome  
0         1     -1       3.0  
1         1    339       0.0  
2         1    330       0.0  
3         4     -1       3.0  
4         1     -1       3.0

In [119]:
data.dtypes

age            int64
job          float64
marital      float64
education    float64
default      float64
balance        int64
housing      float64
loan         float64
day            int64
month        float64
campaign       int64
pdays          int64
poutcome     float64
dtype: object

### Exercise 4

**Perform the data split using the provided SKLearn function on the encoded dataset.**

In [120]:
from sklearn.model_selection import train_test_split

In [121]:
X_train, X_test, y_train, y_test = train_test_split(data, target, random_state=42)

### Exercise 5

**Select a simple supervised model to train with the dataset and explain the decision.**

In [122]:
from sklearn.tree import DecisionTreeClassifier, plot_tree

I decided to start off with a fairly simple model with good performance but still great explainability. I chose to use the DecisionTree for this classification problem.

In [123]:
tree = DecisionTreeClassifier()
tree.fit(X_train, y_train)

print("DecisionTreeClassifier training score: {:.3f}".format(tree.score(X_train,y_train)))
print("DecisionTreeClassifier test score: {:.3f}".format(tree.score(X_test,y_test)))

DecisionTreeClassifier training score: 1.000
DecisionTreeClassifier test score: 0.823


The tree needs some pruning because it is overfitting. We can see that by comparing its performance on the training dataset to the test dataset.

In [124]:
display(pd.DataFrame([tree.feature_importances_],columns=list(data.columns)))

age       job   marital  education   default   balance   housing  \
0  0.146443  0.076891  0.023783     0.0385  0.007204  0.228335  0.022919   

       loan       day     month  campaign     pdays  poutcome  
0  0.007525  0.153829  0.101437   0.06242  0.075266  0.055448

### Exercise 6

**Train a GradientBoostingClassifier, with n_estimators=10 and learning_rate=1, and compare the results with the previous algorithm.**

In [125]:
from sklearn.ensemble import GradientBoostingClassifier

In [126]:
ens_boost = GradientBoostingClassifier(n_estimators=10, learning_rate=1)
ens_boost.fit(X_train, y_train)

GradientBoostingClassifier(learning_rate=1, n_estimators=10)

In [127]:
ens_boost.score(X_train, y_train)

0.9047197640117994

In [128]:
ens_boost.score(X_test, y_test)

0.8815207780725022

The GradientBoostingClassifier performs significantly better than the DecisionTree, obtaining a score of .882 vs the tree's .83

In [129]:
from sklearn.metrics import confusion_matrix
y_pred_ens_boost = ens_boost.predict(X_test)
confusion_matrix(y_test, y_pred_ens_boost)

array([[966,  40],
       [ 94,  31]], dtype=int64)

In [130]:
display(pd.DataFrame([ens_boost.feature_importances_],columns=data.columns.values.tolist()))

age       job   marital  education  default   balance   housing  \
0  0.239982  0.042621  0.009432   0.007348      0.0  0.134442  0.033294   

      loan       day     month  campaign     pdays  poutcome  
0  0.00537  0.077271  0.215642       0.0  0.127398    0.1072

### Exercise 7

**Using the PolynomialFeatures class, compute the second degree interactions of the dataset.** 

**Retrain the original model (selected by the student in question 5) and compare the results.**


In [131]:
from sklearn.preprocessing import PolynomialFeatures
pol = PolynomialFeatures(degree=2)
X_poly = pol.fit_transform(data)
X_train_p, X_test_p, y_train_p, y_test_p = train_test_split(X_poly, target, random_state=42)

In [132]:
tree2 = DecisionTreeClassifier()
tree2.fit(X_train, y_train)

print("DecisionTreeClassifier training score: {:.3f}".format(tree2.score(X_train,y_train)))
print("DecisionTreeClassifier test score: {:.3f}".format(tree2.score(X_test,y_test)))

DecisionTreeClassifier training score: 1.000
DecisionTreeClassifier test score: 0.825


This method did not make the DecisionTree perform any better. In fact, it performed worse and it still has a problem with overfitting which would be dealt with some pruning.

### Exercise 8

__Using the function cross_val_score, and a Kfold with 5 folds, validate the original trained model (selected by the student in question 5) and the GradientBoostingClassifier.__

__Which conclusions can be extracted?__

In [133]:
from sklearn.model_selection import KFold, cross_val_score, cross_validate
kfold = KFold(n_splits=5)
val = cross_validate(tree, data, target, cv=kfold)
print(val)
val_score = cross_val_score(tree, data, target, cv=kfold)
print("Mean validation scores: {:.3f}".format(val_score.mean()))

from sklearn.model_selection import KFold, cross_val_score, cross_validate
kfold = KFold(n_splits=5)
val = cross_validate(ens_boost, data, target, cv=kfold)
print(val)
val_score = cross_val_score(ens_boost, data, target, cv=kfold)
print("Mean validation scores: {:.3f}".format(val_score.mean()))

{'fit_time': array([0.01600385, 0.01400232, 0.0150032 , 0.01500273, 0.01400256]), 'score_time': array([0.00100088, 0.00200057, 0.0010004 , 0.00100017, 0.00200081]), 'test_score': array([0.79447514, 0.8130531 , 0.81637168, 0.83075221, 0.81084071])}
Mean validation scores: 0.819
{'fit_time': array([0.03900361, 0.04100776, 0.05701232, 0.04000926, 0.04300976]), 'score_time': array([0.00200105, 0.00300097, 0.00400114, 0.00200009, 0.00398397]), 'test_score': array([0.88950276, 0.86836283, 0.88827434, 0.88053097, 0.87831858])}
Mean validation scores: 0.881


We can conclude that, utilizing the same validation technique, the GradientBoostingClassifier performs better than the more simple DecisionTree. Since both are easily interpretable, if we were to put one into production it would be the better performing GradientBoostingClassifier.

### Exercise 9

**Using the original (not encoded) dataset, apply the OneHotEncoder transformation to the whole dataset.**

**Is there any difference in the number of features/columns than in the original?**

In [134]:
data2 = pd.read_csv('C:/Users/Asus/Desktop/Python Projects/AI 2 Course/bank.csv', delimiter=';')
data2.head()

age          job  marital  education default  balance housing loan  \
0   30   unemployed  married    primary      no     1787      no   no   
1   33     services  married  secondary      no     4789     yes  yes   
2   35   management   single   tertiary      no     1350     yes   no   
3   30   management  married   tertiary      no     1476     yes  yes   
4   59  blue-collar  married  secondary      no        0     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome   y  
0  cellular   19   oct        79         1     -1         0  unknown  no  
1  cellular   11   may       220         1    339         4  failure  no  
2  cellular   16   apr       185         1    330         1  failure  no  
3   unknown    3   jun       199         4     -1         0  unknown  no  
4   unknown    5   may       226         1     -1         0  unknown  no

In [135]:
dt = pd.read_csv('C:/Users/Asus/Desktop/Python Projects/AI 2 Course/bank.csv', delimiter=';')
dt.head()
tgt = pd.DataFrame(data=dt['y'])
dt.drop(['y'], axis=1, inplace=True)

In [136]:
from sklearn.preprocessing import OneHotEncoder

In [137]:
enc_o = OneHotEncoder(sparse=False)
enc_o.fit(dt)
d_trans = enc_o.transform(dt)
print(dt.to_numpy().shape)
print(d_trans.shape)

(4521, 16)
(4521, 3718)


In [138]:
enc_o.fit(tgt)
t_trans = enc_o.transform(tgt)

print(tgt.to_numpy().shape)
print(t_trans.shape)

(4521, 1)
(4521, 2)


Yes, there is a big difference in the number of features/columns than in the original. This happens because the OneHotEncoder creates one column for each category in each categorical variable or each number in each numerical variable.

Usually we wouldn't OneHotEncode numerical variables because it removes its magnitude/proportions

### Exercise 10

**Retrain the original model (selected by the student in question 5) using the new encoded dataset, and compare the results with the previous results.**

**Validate this model with this dataset using the same functions as in question 8.**

In [139]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import KFold, cross_val_score, cross_validate

In [140]:
X_train2, X_test2, y_train2, y_test2 = train_test_split(d_trans, t_trans, random_state=42)

tree2 = DecisionTreeClassifier()
tree2.fit(X_train2, y_train2)

print("DecisionTreeClassifier training score: {:.3f}".format(tree2.score(X_train2,y_train2)))
print("DecisionTreeClassifier test score: {:.3f}".format(tree2.score(X_test2,y_test2)))

DecisionTreeClassifier training score: 1.000
DecisionTreeClassifier test score: 0.878


In [141]:
kfold = KFold(n_splits=5)
val = cross_validate(tree2, data, target, cv=kfold)
print(val)
val_score = cross_val_score(tree2, data, target, cv=kfold)
print("Mean validation scores: {:.3f}".format(val_score.mean()))

{'fit_time': array([0.01899648, 0.01499009, 0.01701617, 0.0250082 , 0.01499772]), 'score_time': array([0.00101256, 0.00200129, 0.0010016 , 0.00200033, 0.00100017]), 'test_score': array([0.80773481, 0.80973451, 0.82411504, 0.82632743, 0.8119469 ])}
Mean validation scores: 0.818


The model ends up performing worse with the OneHotEncoding than before. This might be due to the reasons stated above in the last question related to the OneHotEncoding of numerical variables which reduces the information obtained through them and consequently might reduce the model's performance.

If we had applied the OneHotEncoding technique only to the categorical variables we might have gotten better scores on the cross-validation.